src/proxy.rs

```diff
@@ -526,6 +530,61 @@ pub fn guess_inbound_service(
         .map(ServiceDescription::from)
 }
 
+// Checks that the source identiy and address match the upstream's waypoint
+async fn check_from_waypoint(
+    state: DemandProxyState,
+    upstream: &Workload,
+    src_identity: Option<&Identity>,
+    src_ip: &IpAddr,
+) -> bool {
+    let is_waypoint = |wl: &Workload| {
+        Some(wl.identity()).as_ref() == src_identity && wl.workload_ips.contains(src_ip)
```

Comment:
Long term it would be nice to have a method more reliable than source IP. That being said, since its node-local it would be quite strange for this check to not work somehow.